File: ui-ngx/src/app/core/http/entity-group.service.ts
///
/// ThingsBoard, Inc. ("COMPANY") CONFIDENTIAL
///
/// Copyright © 2016-2023 ThingsBoard, Inc. All Rights Reserved.
///
/// NOTICE: All information contained herein is, and remains
/// the property of ThingsBoard, Inc. and its suppliers,
/// if any.  The intellectual and technical concepts contained
/// herein are proprietary to ThingsBoard, Inc.
/// and its suppliers and may be covered by U.S. and Foreign Patents,
/// patents in process, and are protected by trade secret or copyright law.
///
/// Dissemination of this information or reproduction of this material is strictly forbidden
/// unless prior written permission is obtained from COMPANY.
///
/// Access to the source code contained herein is hereby forbidden to anyone except current COMPANY employees,
/// managers or contractors who have executed Confidentiality and Non-disclosure agreements
/// explicitly covering such access.
///
/// The copyright notice above does not evidence any actual or intended publication
/// or disclosure  of  this source code, which includes
/// information that is confidential and/or proprietary, and is a trade secret, of  COMPANY.
/// ANY REPRODUCTION, MODIFICATION, DISTRIBUTION, PUBLIC  PERFORMANCE,
/// OR PUBLIC DISPLAY OF OR THROUGH USE  OF THIS  SOURCE CODE  WITHOUT
/// THE EXPRESS WRITTEN CONSENT OF COMPANY IS STRICTLY PROHIBITED,
/// AND IN VIOLATION OF APPLICABLE LAWS AND INTERNATIONAL TREATIES.
/// THE RECEIPT OR POSSESSION OF THIS SOURCE CODE AND/OR RELATED INFORMATION
/// DOES NOT CONVEY OR IMPLY ANY RIGHTS TO REPRODUCE, DISCLOSE OR DISTRIBUTE ITS CONTENTS,
/// OR TO MANUFACTURE, USE, OR SELL ANYTHING THAT IT  MAY DESCRIBE, IN WHOLE OR IN PART.
///

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { PageLink } from '@shared/models/page/page-link';
import { defaultHttpOptionsFromConfig, RequestConfig } from '@core/http/http-utils';
import { EMPTY, forkJoin, Observable, of, throwError } from 'rxjs';
import { PageData } from '@shared/models/page/page-data';
import { ContactBased } from '@shared/models/contact-based.model';
import { EntityId } from '@shared/models/id/entity-id';
import {
  DeviceEntityGroupInfo,
  EntityGroup,
  EntityGroupInfo,
  prepareEntityGroupConfiguration,
  ShareGroupRequest,
  ShortEntityView
} from '@shared/models/entity-group.models';
import { EntityType } from '@shared/models/entity-type.models';
import { EntityGroupId } from '@shared/models/id/entity-group-id';
import { concatMap, expand, map, mergeMap, toArray } from 'rxjs/operators';
import { BaseData, HasId, sortEntitiesByIds } from '@shared/models/base-data';
import { deepClone, isDefinedAndNotNull } from '@core/utils';
import { OtaPackageService } from '@core/http/ota-package.service';
import { DeviceGroupOtaPackage, OtaUpdateType } from '@shared/models/ota-package.models';
import { OtaPackageId } from '@shared/models/id/ota-package-id';
import { Direction } from '@shared/models/page/sort-order';
import { EntityInfo, EntityInfoData } from '@shared/models/entity.models';

@Injectable({
  providedIn: 'root'
})
export class EntityGroupService {

  constructor(
    private http: HttpClient,
    private otaPackageService: OtaPackageService
  ) {}

  public getOwners(pageLink: PageLink, config?: RequestConfig): Observable<PageData<ContactBased<EntityId>>> {
    return this.http.get<PageData<ContactBased<EntityId>>>(`/api/owners${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getOwnerInfos(pageLink: PageLink, config?: RequestConfig): Observable<PageData<EntityInfoData>> {
    return this.http.get<PageData<EntityInfoData>>(`/api/ownerInfos${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getOwnerInfo(ownerId: EntityId, config?: RequestConfig): Observable<EntityInfoData> {
    return this.http.get<EntityInfoData>(`/api/ownerInfo/${ownerId.entityType}/${ownerId.id}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroup(entityGroupId: string, config?: RequestConfig): Observable<EntityGroupInfo> {
    return this.http.get<EntityGroupInfo>(`/api/entityGroup/${entityGroupId}`,
      defaultHttpOptionsFromConfig(config)).pipe(
      map(group => {
        group.configuration = prepareEntityGroupConfiguration(group.type, group.configuration);
        return group;
      })
    );
  }

  public getDeviceEntityGroup(entityGroupId: string, config?: RequestConfig): Observable<DeviceEntityGroupInfo> {
    return this.http.get<EntityGroupInfo>(`/api/entityGroup/${entityGroupId}`,
      defaultHttpOptionsFromConfig(config)).pipe(
      map(group => {
        group.configuration = prepareEntityGroupConfiguration(group.type, group.configuration);
        return group;
      }),
      mergeMap(group => this.fetchOtaPackageGroupInfo(group, config))
    );
  }

  public fetchOtaPackageGroupInfo(group: EntityGroupInfo, config?: RequestConfig): Observable<DeviceEntityGroupInfo> {
    if (isDefinedAndNotNull(group) && group.type === EntityType.DEVICE && !group.groupAll) {
      const tasks = [];
      tasks.push(this.otaPackageService.getOtaPackageInfoByDeviceGroupId(group.id.id, OtaUpdateType.FIRMWARE, config));
      tasks.push(this.otaPackageService.getOtaPackageInfoByDeviceGroupId(group.id.id, OtaUpdateType.SOFTWARE, config));
      return forkJoin(tasks).pipe(
        map(([firmware, software]: DeviceGroupOtaPackage[]) => {
          const deviceGroup = group as DeviceEntityGroupInfo;
          if (isDefinedAndNotNull(firmware)) {
            deviceGroup.firmwareGroup = firmware;
            deviceGroup.firmwareId = deepClone(firmware.otaPackageId);
          }
          if (isDefinedAndNotNull(software)) {
            deviceGroup.softwareGroup = software;
            deviceGroup.softwareId = deepClone(software.otaPackageId);
          }
          return deviceGroup;
        })
      );
    }
    return of(group);
  }

  public updateDeviceGroupOtaPackage(deviceGroupOtaPackage: DeviceGroupOtaPackage, otaPackageId: OtaPackageId | null,
                                     groupId: EntityGroupId, otaPackageType: OtaUpdateType,
                                     config?: RequestConfig): Observable<DeviceGroupOtaPackage> {
    if (isDefinedAndNotNull(deviceGroupOtaPackage)) {
      if (otaPackageId === null) {
        return this.deleteDeviceGroupOtaPackage(deviceGroupOtaPackage.id, config);
      } else if (otaPackageId.id !== deviceGroupOtaPackage.otaPackageId.id) {
        deviceGroupOtaPackage.otaPackageId = otaPackageId;
        return this.saveDeviceGroupOtaPackage(deviceGroupOtaPackage, config);
      } else {
        return of(deviceGroupOtaPackage);
      }
    } else if (isDefinedAndNotNull(otaPackageId)) {
      const groupOtaPackage: DeviceGroupOtaPackage = {
        otaPackageId,
        otaPackageType,
        groupId
      };
      return this.saveDeviceGroupOtaPackage(groupOtaPackage, config);
    }
    return of(null);
  }

  public saveDeviceGroupOtaPackage(deviceGroupOtaPackage: DeviceGroupOtaPackage,
                                   config?: RequestConfig ): Observable<DeviceGroupOtaPackage> {
    return this.http.post<DeviceGroupOtaPackage>('/api/deviceGroupOtaPackage', deviceGroupOtaPackage, defaultHttpOptionsFromConfig(config));
  }

  public saveDeviceEntityGroup(entityGroup: DeviceEntityGroupInfo, originalEntityGroup: DeviceEntityGroupInfo,
                               config?: RequestConfig): Observable<DeviceEntityGroupInfo> {
    if (isDefinedAndNotNull(entityGroup.id)) {
      return this.otaPackageService.confirmDialogUpdatePackage(entityGroup, originalEntityGroup).pipe(
        mergeMap((update) => {
          if (update) {
            const firmwareId = entityGroup.firmwareId;
            const firmwareGroup = entityGroup.firmwareGroup;
            const softwareId = entityGroup.softwareId;
            const softwareGroup = entityGroup.softwareGroup;
            delete entityGroup.firmwareId;
            delete entityGroup.firmwareGroup;
            delete entityGroup.softwareId;
            delete entityGroup.softwareGroup;
            const tasks = [this.updateDeviceGroupOtaPackage(firmwareGroup, firmwareId, entityGroup.id, OtaUpdateType.FIRMWARE, config),
                           this.updateDeviceGroupOtaPackage(softwareGroup, softwareId, entityGroup.id, OtaUpdateType.SOFTWARE, config),
                           this.saveEntityGroup(entityGroup, config)];
            return forkJoin(tasks).pipe(
              map(([firmware, software, savedEntityGroup]: [DeviceGroupOtaPackage, DeviceGroupOtaPackage, EntityGroupInfo]) =>
                Object.assign(savedEntityGroup, {
                    firmwareId: deepClone(firmware?.otaPackageId),
                    firmwareGroup: firmware,
                    softwareId: deepClone(software?.otaPackageId),
                    softwareGroup: software
                  })
              ));
          }
          return throwError(() => 'Canceled saving device group');
        })
      );
    }
    return this.saveEntityGroup(entityGroup, config);
  }

  public deleteDeviceGroupOtaPackage(otaPackageId: string, config?: RequestConfig ) {
    return this.http.delete<null>(`/api/deviceGroupOtaPackage/${otaPackageId}`, defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupEntityInfo(entityGroupId: string, config?: RequestConfig): Observable<EntityInfoData> {
    return this.http.get<EntityInfoData>(`/api/entityGroupInfo/${entityGroupId}`,
      defaultHttpOptionsFromConfig(config));
  }

  public saveEntityGroup(entityGroup: EntityGroup, config?: RequestConfig): Observable<EntityGroupInfo> {
    return this.http.post<EntityGroupInfo>('/api/entityGroup', entityGroup, defaultHttpOptionsFromConfig(config));
  }

  public deleteEntityGroup(entityGroupId: string, config?: RequestConfig) {
    return this.http.delete(`/api/entityGroup/${entityGroupId}`, defaultHttpOptionsFromConfig(config));
  }

  public makeEntityGroupPublic(entityGroupId: string, config?: RequestConfig): Observable<any> {
    return this.http.post(`/api/entityGroup/${entityGroupId}/makePublic`, null, defaultHttpOptionsFromConfig(config));
  }

  public makeEntityGroupPrivate(entityGroupId: string, config?: RequestConfig): Observable<any> {
    return this.http.post(`/api/entityGroup/${entityGroupId}/makePrivate`, null, defaultHttpOptionsFromConfig(config));
  }

  public shareEntityGroup(entityGroupId: string, shareGroupRequest: ShareGroupRequest, config?: RequestConfig): Observable<any> {
    return this.http.post(`/api/entityGroup/${entityGroupId}/share`, shareGroupRequest, defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroups(pageLink: PageLink, groupType: EntityType,
                         includeShared = true, config?: RequestConfig): Observable<PageData<EntityGroupInfo>> {
    return this.http.get<PageData<EntityGroupInfo>>(`/api/entityGroups/${groupType}${pageLink.toQuery()}&includeShared=${includeShared}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupEntityInfos(pageLink: PageLink, groupType: EntityType,
                                   includeShared = true, config?: RequestConfig): Observable<PageData<EntityInfoData>> {
    return this.http.get<PageData<EntityInfoData>>(`/api/entityGroupInfos/${groupType}${pageLink.toQuery()}&includeShared=${includeShared}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getAllEntityGroups(pageSize: number, groupType: EntityType, includeShared = true,
                            config?: RequestConfig): Observable<Array<EntityGroupInfo>> {
    const pageLink = new PageLink(pageSize, 0, null, {
      property: 'name',
      direction: Direction.ASC
    });
    if (pageSize === -1) { // all
      pageLink.pageSize = groupType === EntityType.CUSTOMER ? 1024 : 100;
      return this.getAllEntityGroupsByPageLink(pageLink, groupType, includeShared, config).pipe(
        map((data) => data && data.length ? data : null)
      );
    } else {
      const entityGroupsObservable: Observable<PageData<EntityGroupInfo>> =
        this.getEntityGroups(pageLink, groupType, includeShared, config);
      if (entityGroupsObservable) {
        return entityGroupsObservable.pipe(
          map((data) => data && data.data.length ? data.data : null)
        );
      } else {
        return of(null);
      }
    }
  }

  private getAllEntityGroupsByPageLink(pageLink: PageLink, groupType: EntityType, includeShared = true,
                                       config?: RequestConfig): Observable<Array<EntityGroupInfo>> {
    const entityGroupsObservable: Observable<PageData<EntityGroupInfo>> =
      this.getEntityGroups(pageLink, groupType, includeShared, config);
    if (entityGroupsObservable) {
      return entityGroupsObservable.pipe(
        expand((data) => {
          if (data.hasNext) {
            pageLink.page += 1;
            return this.getEntityGroups(pageLink, groupType, includeShared, config);
          } else {
            return EMPTY;
          }
        }),
        map((data) => data.data),
        concatMap((data) => data),
        toArray()
      );
    } else {
      return of(null);
    }
  }

  public getSharedEntityGroups(pageLink: PageLink, groupType: EntityType, config?: RequestConfig): Observable<PageData<EntityGroupInfo>> {
    return this.http.get<PageData<EntityGroupInfo>>(`/api/entityGroups/${groupType}/shared${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getSharedEntityGroupEntityInfos(pageLink: PageLink, groupType: EntityType,
                                         config?: RequestConfig): Observable<PageData<EntityInfoData>> {
    return this.http.get<PageData<EntityInfoData>>(`/api/entityGroupInfos/${groupType}/shared${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupIdsForEntityId(entityType: EntityType, entityId: string, config?: RequestConfig): Observable<Array<EntityGroupId>> {
    return this.http.get<Array<EntityGroupId>>(`/api/entityGroups/${entityType}/${entityId}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupsByIds(entityGroupIds: Array<string>, config?: RequestConfig): Observable<Array<EntityGroupInfo>> {
    return this.http.get<Array<EntityGroupInfo>>(`/api/entityGroups?entityGroupIds=${entityGroupIds.join(',')}`,
      defaultHttpOptionsFromConfig(config)).pipe(
      map((entityGroups) => sortEntitiesByIds(entityGroups, entityGroupIds))
    );
  }

  public getEntityGroupEntityInfosByIds(entityGroupIds: Array<string>, config?: RequestConfig): Observable<Array<EntityInfoData>> {
    return this.http.get<Array<EntityInfoData>>(`/api/entityGroupInfos?entityGroupIds=${entityGroupIds.join(',')}`,
      defaultHttpOptionsFromConfig(config)).pipe(
      map((entityGroups) => sortEntitiesByIds(entityGroups, entityGroupIds))
    );
  }

  public getEntityGroupsHierarchyByOwnerId(pageLink: PageLink, ownerType: EntityType, ownerId: string, groupType: EntityType,
                                           config?: RequestConfig): Observable<PageData<EntityGroupInfo>> {
    return this.http.get<PageData<EntityGroupInfo>>(`/api/entityGroupsHierarchy/${ownerType}/${ownerId}/${groupType}${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupEntityInfosHierarchyByOwnerId(pageLink: PageLink, ownerType: EntityType, ownerId: string, groupType: EntityType,
                                                     config?: RequestConfig): Observable<PageData<EntityInfoData>> {
    return this.http.get<PageData<EntityInfoData>>(
      `/api/entityGroupInfosHierarchy/${ownerType}/${ownerId}/${groupType}${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupsByOwnerId(pageLink: PageLink, ownerType: EntityType, ownerId: string, groupType: EntityType,
                                  config?: RequestConfig): Observable<PageData<EntityGroupInfo>> {
    return this.http.get<PageData<EntityGroupInfo>>(`/api/entityGroups/${ownerType}/${ownerId}/${groupType}${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupEntityInfosByOwnerId(pageLink: PageLink, ownerType: EntityType, ownerId: string, groupType: EntityType,
                                  config?: RequestConfig): Observable<PageData<EntityInfoData>> {
    return this.http.get<PageData<EntityInfoData>>(`/api/entityGroupInfos/${ownerType}/${ownerId}/${groupType}${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getAllEntityGroupsByOwnerId(pageSize: number, ownerType: EntityType, ownerId: string, groupType: EntityType,
                                     config?: RequestConfig): Observable<Array<EntityGroupInfo>> {
    const pageLink = new PageLink(pageSize, 0, null, {
      property: 'name',
      direction: Direction.ASC
    });
    if (pageSize === -1) { // all
      pageLink.pageSize = groupType === EntityType.CUSTOMER ? 1024 : 100;
      return this.getAllEntityGroupsByOwnerIdAndPageLink(pageLink, ownerType, ownerId, groupType, config).pipe(
        map((data) => data && data.length ? data : null)
      );
    } else {
      const entityGroupsObservable: Observable<PageData<EntityGroupInfo>> =
        this.getEntityGroupsByOwnerId(pageLink, ownerType, ownerId, groupType, config);
      if (entityGroupsObservable) {
        return entityGroupsObservable.pipe(
          map((data) => data && data.data.length ? data.data : null)
        );
      } else {
        return of(null);
      }
    }
  }

  private getAllEntityGroupsByOwnerIdAndPageLink(pageLink: PageLink, ownerType: EntityType, ownerId: string, groupType: EntityType,
                                                 config?: RequestConfig): Observable<Array<EntityGroupInfo>> {
    const entityGroupsObservable: Observable<PageData<EntityGroupInfo>> =
      this.getEntityGroupsByOwnerId(pageLink, ownerType, ownerId, groupType, config);
    if (entityGroupsObservable) {
      return entityGroupsObservable.pipe(
        expand((data) => {
          if (data.hasNext) {
            pageLink.page += 1;
            return this.getEntityGroupsByOwnerId(pageLink, ownerType, ownerId, groupType, config);
          } else {
            return EMPTY;
          }
        }),
        map((data) => data.data),
        concatMap((data) => data),
        toArray()
      );
    } else {
      return of(null);
    }
  }

  public getEntityGroupsByPageLink(pageLink: PageLink, groupType: EntityType,
                                   config?: RequestConfig): Observable<PageData<EntityGroupInfo>> {
    return this.getEntityGroups(pageLink, groupType, true, config);
  }

  public getEntityGroupAllByOwnerId(ownerType: EntityType, ownerId: string, groupType: EntityType,
                                    config?: RequestConfig): Observable<EntityGroupInfo> {
    return this.http.get<EntityGroupInfo>(`/api/entityGroup/all/${ownerType}/${ownerId}/${groupType}`,
      defaultHttpOptionsFromConfig(config));
  }

  public addEntityToEntityGroup(entityGroupId: string, entityId: string, config?: RequestConfig): Observable<any> {
    return this.http.post(`/api/entityGroup/${entityGroupId}/addEntities`, [entityId], defaultHttpOptionsFromConfig(config));
  }

  public addEntitiesToEntityGroup(entityGroupId: string, entityIds: string[], config?: RequestConfig): Observable<any> {
    return this.http.post(`/api/entityGroup/${entityGroupId}/addEntities`, entityIds, defaultHttpOptionsFromConfig(config));
  }

  public changeEntityOwner(ownerId: EntityId, entityId: EntityId, entityGroupIds?: string[], config?: RequestConfig): Observable<any> {
    return this.http.post(`/api/owner/${ownerId.entityType}/${ownerId.id}/${entityId.entityType}/${entityId.id}`,
      entityGroupIds, defaultHttpOptionsFromConfig(config));
  }

  public removeEntityFromEntityGroup(entityGroupId: string, entityId: string, config?: RequestConfig): Observable<any> {
    return this.http.post(`/api/entityGroup/${entityGroupId}/deleteEntities`, [entityId], defaultHttpOptionsFromConfig(config));
  }

  public removeEntitiesFromEntityGroup(entityGroupId: string, entityIds: string[], config?: RequestConfig): Observable<any> {
    return this.http.post(`/api/entityGroup/${entityGroupId}/deleteEntities`, entityIds, defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupEntity(entityGroupId: string, entityId: string, config?: RequestConfig): Observable<ShortEntityView> {
    return this.http.get<ShortEntityView>(`/api/entityGroup/${entityGroupId}/${entityId}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getEntityGroupEntities<T extends BaseData<HasId> | ShortEntityView>(entityGroupId: string,
                                                                             pageLink: PageLink,
                                                                             entityType?: EntityType,
                                                                             config?: RequestConfig): Observable<PageData<T>> {
    let url = `/api/entityGroup/${entityGroupId}/`;
    if (entityType) {
      switch (entityType) {
        case EntityType.ASSET:
          url += 'assets';
          break;
        case EntityType.CUSTOMER:
          url += 'customers';
          break;
        case EntityType.DEVICE:
          url += 'devices';
          break;
        case EntityType.USER:
          url += 'users';
          break;
        case EntityType.ENTITY_VIEW:
          url += 'entityViews';
          break;
        case EntityType.DASHBOARD:
          url += 'dashboards';
          break;
        case EntityType.EDGE:
          url += 'edges';
          break;
        default:
          url += 'entities';
      }
    } else {
      url += 'entities';
    }
    url += pageLink.toQuery();
    return this.http.get<PageData<T>>(url, defaultHttpOptionsFromConfig(config));
  }

  public getEdgeEntityGroups(pageLink: PageLink, edgeId: string, groupType: EntityType,
                             config?: RequestConfig): Observable<PageData<EntityGroupInfo>> {
    return this.http.get<PageData<EntityGroupInfo>>(`/api/entityGroups/edge/${edgeId}/${groupType}${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getAllEdgeEntityGroups(pageSize: number, edgeId: string, groupType: EntityType,
                                config?: RequestConfig): Observable<Array<EntityGroupInfo>> {
    const pageLink = new PageLink(pageSize, 0, null, {
      property: 'name',
      direction: Direction.ASC
    });
    if (pageSize === -1) { // all
      pageLink.pageSize = groupType === EntityType.CUSTOMER ? 1024 : 100;
      return this.getAllEdgeEntityGroupsByPageLink(pageLink, edgeId, groupType, config).pipe(
        map((data) => data && data.length ? data : null)
      );
    } else {
      const entityGroupsObservable: Observable<PageData<EntityGroupInfo>> =
        this.getEdgeEntityGroups(pageLink, edgeId, groupType, config);
      if (entityGroupsObservable) {
        return entityGroupsObservable.pipe(
          map((data) => data && data.data.length ? data.data : null)
        );
      } else {
        return of(null);
      }
    }
  }

  private getAllEdgeEntityGroupsByPageLink(pageLink: PageLink, edgeId: string, groupType: EntityType,
                                           config?: RequestConfig): Observable<Array<EntityGroupInfo>> {
    const entityGroupsObservable: Observable<PageData<EntityGroupInfo>> =
      this.getEdgeEntityGroups(pageLink, edgeId, groupType, config);
    if (entityGroupsObservable) {
      return entityGroupsObservable.pipe(
        expand((data) => {
          if (data.hasNext) {
            pageLink.page += 1;
            return this.getEdgeEntityGroups(pageLink, edgeId, groupType, config);
          } else {
            return EMPTY;
          }
        }),
        map((data) => data.data),
        concatMap((data) => data),
        toArray()
      );
    } else {
      return of(null);
    }
  }

  public assignEntityGroupToEdge(edgeId: string, entityGroupId: string,
                                 groupType: string, config?: RequestConfig): Observable<EntityGroup> {
    return this.http.post<EntityGroup>(`/api/edge/${edgeId}/entityGroup/${entityGroupId}/${groupType}`,
      defaultHttpOptionsFromConfig(config));
  }

  public unassignEntityGroupFromEdge(edgeId: string, entityGroupId: string,
                                     groupType: string, config?: RequestConfig): Observable<EntityGroup> {
    return this.http.delete<EntityGroup>(`/api/edge/${edgeId}/entityGroup/${entityGroupId}/${groupType}`,
      defaultHttpOptionsFromConfig(config));
  }
}

File: common/dao-api/src/main/java/org/thingsboard/server/dao/group/EntityGroupService.java
/**
 * ThingsBoard, Inc. ("COMPANY") CONFIDENTIAL
 *
 * Copyright © 2016-2023 ThingsBoard, Inc. All Rights Reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of ThingsBoard, Inc. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to ThingsBoard, Inc.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 *
 * Dissemination of this information or reproduction of this material is strictly forbidden
 * unless prior written permission is obtained from COMPANY.
 *
 * Access to the source code contained herein is hereby forbidden to anyone except current COMPANY employees,
 * managers or contractors who have executed Confidentiality and Non-disclosure agreements
 * explicitly covering such access.
 *
 * The copyright notice above does not evidence any actual or intended publication
 * or disclosure  of  this source code, which includes
 * information that is confidential and/or proprietary, and is a trade secret, of  COMPANY.
 * ANY REPRODUCTION, MODIFICATION, DISTRIBUTION, PUBLIC  PERFORMANCE,
 * OR PUBLIC DISPLAY OF OR THROUGH USE  OF THIS  SOURCE CODE  WITHOUT
 * THE EXPRESS WRITTEN CONSENT OF COMPANY IS STRICTLY PROHIBITED,
 * AND IN VIOLATION OF APPLICABLE LAWS AND INTERNATIONAL TREATIES.
 * THE RECEIPT OR POSSESSION OF THIS SOURCE CODE AND/OR RELATED INFORMATION
 * DOES NOT CONVEY OR IMPLY ANY RIGHTS TO REPRODUCE, DISCLOSE OR DISTRIBUTE ITS CONTENTS,
 * OR TO MANUFACTURE, USE, OR SELL ANYTHING THAT IT  MAY DESCRIBE, IN WHOLE OR IN PART.
 */
package org.thingsboard.server.dao.group;

import com.google.common.util.concurrent.ListenableFuture;
import org.thingsboard.server.common.data.EntityType;
import org.thingsboard.server.common.data.ShortEntityView;
import org.thingsboard.server.common.data.edge.Edge;
import org.thingsboard.server.common.data.group.EntityGroup;
import org.thingsboard.server.common.data.group.EntityGroupInfo;
import org.thingsboard.server.common.data.id.CustomerId;
import org.thingsboard.server.common.data.id.EdgeId;
import org.thingsboard.server.common.data.id.EntityGroupId;
import org.thingsboard.server.common.data.id.EntityId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.page.PageData;
import org.thingsboard.server.common.data.page.PageLink;
import org.thingsboard.server.common.data.permission.MergedUserPermissions;
import org.thingsboard.server.dao.entity.EntityDaoService;

import java.util.List;
import java.util.Optional;

public interface EntityGroupService extends EntityDaoService {

    EntityGroup findEntityGroupById(TenantId tenantId, EntityGroupId entityGroupId);

    EntityGroupInfo findEntityGroupInfoById(TenantId tenantId, EntityGroupId entityGroupId);

    ListenableFuture<EntityGroup> findEntityGroupByIdAsync(TenantId tenantId, EntityGroupId entityGroupId);

    ListenableFuture<EntityGroupInfo> findEntityGroupInfoByIdAsync(TenantId tenantId, EntityGroupId entityGroupId);

    EntityGroup saveEntityGroup(TenantId tenantId, EntityId parentEntityId, EntityGroup entityGroup);

    EntityGroup createEntityGroupAll(TenantId tenantId, EntityId parentEntityId, EntityType groupType);

    EntityGroup findOrCreateUserGroup(TenantId tenantId, EntityId parentEntityId, String groupName, String description);

    EntityGroup findOrCreateEntityGroup(TenantId tenantId, EntityId parentEntityId, EntityType groupType, String groupName,
                                        String description, CustomerId publicCustomerId);

    Optional<EntityGroupInfo> findOwnerEntityGroupInfo(TenantId tenantId, EntityId parentEntityId, EntityType groupType, String groupName);

    EntityGroup findOrCreateTenantUsersGroup(TenantId tenantId);

    EntityGroup findOrCreateTenantAdminsGroup(TenantId tenantId);

    EntityGroup findOrCreateCustomerUsersGroup(TenantId tenantId, CustomerId customerId, CustomerId parentCustomerId);

    EntityGroup findOrCreateCustomerAdminsGroup(TenantId tenantId, CustomerId customerId, CustomerId parentCustomerId);

    EntityGroup findOrCreatePublicUsersGroup(TenantId tenantId, CustomerId customerId);

    EntityGroup findOrCreateReadOnlyEntityGroupForCustomer(TenantId tenantId, CustomerId customerId, EntityType groupType);

    ListenableFuture<Optional<EntityGroup>> findPublicUserGroupAsync(TenantId tenantId, CustomerId publicCustomerId);

    void deleteEntityGroup(TenantId tenantId, EntityGroupId entityGroupId);

    PageData<EntityGroup> findAllEntityGroupsByParentRelation(TenantId tenantId, EntityId parentEntityId, PageLink pageLink);

    void deleteAllEntityGroups(TenantId tenantId, EntityId parentEntityId);

    PageData<EntityGroup> findEntityGroupsByType(TenantId tenantId, EntityId parentEntityId, EntityType groupType, PageLink pageLink);

    PageData<EntityGroupInfo> findEntityGroupInfosByType(TenantId tenantId, EntityId parentEntityId,
                                                         EntityType groupType, PageLink pageLink);

    PageData<EntityGroupInfo> findEntityGroupInfosByOwnersAndType(TenantId tenantId, List<EntityId> ownerIds,
                                                                  EntityType groupType, PageLink pageLink);

    PageData<EntityGroupInfo> findEntityGroupInfosByIds(TenantId tenantId, List<EntityGroupId> entityGroupIds, PageLink pageLink);

    PageData<EntityGroupInfo> findEntityGroupInfosByTypeOrIds(TenantId tenantId, EntityId parentEntityId,
                                                              EntityType groupType, List<EntityGroupId> entityGroupIds, PageLink pageLink);

    PageData<EntityGroupInfo> findEdgeEntityGroupInfosByOwnerIdType(TenantId tenantId, EdgeId edgeId, EntityId ownerId, EntityType groupType, PageLink pageLink);

    PageData<EntityGroup> findEntityGroupsByType(TenantId tenantId, EntityType groupType, PageLink pageLink);

    Optional<EntityGroup> findEntityGroupByTypeAndName(TenantId tenantId, EntityId parentEntityId, EntityType groupType, String name);

    Optional<EntityGroupInfo> findEntityGroupInfoByTypeAndName(TenantId tenantId, EntityId parentEntityId, EntityType groupType, String name);

    ListenableFuture<Optional<EntityGroup>> findEntityGroupByTypeAndNameAsync(TenantId tenantId, EntityId parentEntityId, EntityType groupType, String name);

    void addEntityToEntityGroup(TenantId tenantId, EntityGroupId entityGroupId, EntityId entityId);

    void addEntityToEntityGroupAll(TenantId tenantId, EntityId parentEntityId, EntityId entityId);

    void addEntitiesToEntityGroup(TenantId tenantId, EntityGroupId entityGroupId, List<EntityId> entityIds);

    void removeEntityFromEntityGroup(TenantId tenantId, EntityGroupId entityGroupId, EntityId entityId);

    void removeEntitiesFromEntityGroup(TenantId tenantId, EntityGroupId entityGroupId, List<EntityId> entityIds);

    ShortEntityView findGroupEntity(TenantId tenantId, CustomerId customerId, MergedUserPermissions userPermissions, EntityGroupId entityGroupId, EntityId entityId);

    PageData<ShortEntityView> findGroupEntities(TenantId tenantId, CustomerId customerId, MergedUserPermissions userPermissions, EntityGroupId entityGroupId, PageLink pageLink);

    ListenableFuture<List<EntityId>> findAllEntityIdsAsync(TenantId tenantId, EntityGroupId entityGroupId, PageLink pageLink);

    PageData<EntityId> findEntityIds(TenantId tenantId, EntityType entityType, EntityGroupId entityGroupId, PageLink pageLink);

    ListenableFuture<List<EntityGroupId>> findEntityGroupsForEntityAsync(TenantId tenantId, EntityId entityId);

    boolean isEntityInGroup(TenantId tenantId, EntityId entityId, EntityGroupId entityGroupId);

    EntityGroup assignEntityGroupToEdge(TenantId tenantId, EntityGroupId entityGroupId, EdgeId edgeId, EntityType groupType);

    EntityGroup unassignEntityGroupFromEdge(TenantId tenantId, EntityGroupId entityGroupId, EdgeId edgeId, EntityType groupType);

    PageData<EntityGroup> findEdgeEntityGroupsByType(TenantId tenantId, EdgeId edgeId, EntityType groupType, PageLink pageLink);

    ListenableFuture<Boolean> checkEdgeEntityGroupByIdAsync(TenantId tenantId, EdgeId edgeId, EntityGroupId entityGroupId, EntityType groupType);

    ListenableFuture<EntityGroup> findOrCreateEdgeAllGroupAsync(TenantId tenantId, Edge edge, String edgeName, EntityType groupType);
}
